internlm/monitor/diagnosis.py

```diff
@@ -0,0 +1,272 @@
+            global_rank_uid = self.get_rank_uid()
+            time_info = self.timer.get_all_timer_results(reset=False)
+            batch_count = self.train_state.batch_count
+            if batch_count > 10:
+                with open(self.diagnosis_path, 'a') as f:
+                    with contextlib.redirect_stdout(f):
+                        if batch_count % 2 == 0:
+                            time_list = (global_rank_uid, time_info)
+                            if gpc.get_global_rank() == 0:
+                                all_rank_time_list = [None for _ in range(gpc.get_world_size(ParallelMode.GLOBAL))]
+                            else:
+                                all_rank_time_list = None
+                            dist.gather_object(time_list, all_rank_time_list, dst=0, group=gpc.get_group(ParallelMode.GLOBAL))
+                            if gpc.get_global_rank() == 0:
+                                all_times = {}
+                                for rank_time_info in all_rank_time_list:
+                                    ruid, info = rank_time_info
+                                    for time_tuple in info:
+                                        name, value = time_tuple
+                                        if name not in all_times:
+                                            all_times[name] = [(value, ruid)]
+                                        else:
+                                            all_times[name].append((value, ruid))
+                                                                        
+                                self._print_table(all_times)
+                                            
+                        if gpc.get_global_rank() == 0: 
+                            slow_list = []        
+                            for time_tuple in time_info:
+                                name, value = time_tuple
+                                if batch_count % 2 == 0:        
+                                    avg_val = self.rank_avg_time[name]
+                                    if value >= avg_val * 1:
+                                        slow_list.append((name, value, avg_val))
+                                sum_val = self.rank_avg_time[name] * (batch_count - 10) + value
+                                self.rank_avg_time[name] = round(sum_val / (batch_count - 9), 2)
+                                             
+                            
+                            if batch_count % 2 == 0 and slow_list != []:
+                                print('Cross step dia:')
+                                print('This step is slower than average')
+                                for tuple in slow_list:
+                                    name, value, avg_val = tuple
+                                    print(f'{name} = {value} > avg ({avg_val})')
+                                print()
+                        
+                        if batch_count % 2 == 0:
```

Comment:
batch_count % 2 == 0，  is a very high frequency. It would be more reasonable to make it a configurable parameter.